src/recipe.cpp

```diff
@@ -352,19 +348,28 @@ void recipe::load( const JsonObject &jo, const std::string &src )
 
 void recipe::finalize()
 {
-    if( test_mode && check_blueprint_needs ) {
+    if( bp_autocalc ) {
+        blueprint_reqs_ = get_build_reqs_for_furn_ter_ids(
+                              get_changed_ids_from_update( blueprint ) );
+    } else if( test_mode && check_blueprint_needs ) {
         check_blueprint_requirements();
     }
 
+    incorporate_build_reqs();
+
     // concatenate both external and inline requirements
     add_requirements( reqs_external );
     add_requirements( reqs_internal );
 
     reqs_external.clear();
     reqs_internal.clear();
 
-    if( has_blueprint_needs ) {
-        blueprint_reqs.clear();
+    if( blueprint_reqs_ ) {
+        blueprint_reqs_.reset();
+    }
+
+    if( bp_autocalc ) {
+        requirements_.consolidate();
     }
```

Comment:
This should either be done for both auto calculated and explicit requirements, or not at all. Otherwise the final requirements may differ when bluepruint requirements are auto calculated or explicitly specified.

Anyway, I suggest we do not consolidate the requirements here, so we don't mix up requirements from different sources.

Reply:
It's necessary to consolidate the autocalculated requirements.  They're too unwieldy otherwise and cause errors in the requirement deduplication step.  This was how it was done before in the previous implementation.  The final requirements won't differ because they're still verified to be the same.  (I confess though, that I don't actually know what consolidate does exactly).

Reply:
Before this PR the autocalculated blueprint requirements were consolidated, and compared against explicitly specified requirements, and then merged with recipe requirements (which usually isn't specified for blueprint recipes, but still can be) without consolidating them again. Now in this PR autocalculated requirements are first merged with recipe requirements and then consolidated (which is the most original behavior, like you said). Since consolidate actually changes the exact meaning of the requirements, this could lead to different behavior when a recipe is autocalculated or explicitly specified, if the said recipe contains recipe requirements in addition to the blueprint one. I have no strong preference for either way of consolidating, but it should be one of them to ensure consistency.

> The final requirements won't differ because they're still verified to be the same.
Your code currently skips the check if `bp_autocalc` is true (and that check only checks blueprint requirements, not the final requirements).

> what consolidate does exactly
There's some detailed explanation in the consolidate code. (I changed it to be more symmetric and defined, before that it returned different results when requirements are in different order, etc.)